master/pkg/stream/stream.go

```diff
@@ -1,32 +1,67 @@
 package stream
 
 import (
+	"encoding/json"
 	"sync"
-
-	"github.com/gorilla/websocket"
 )
 
 // Msg is an object with a message and a sequence number and json marshal caching.
 type Msg interface {
 	SeqNum() int64
-	UpsertMsg() *websocket.PreparedMessage
-	DeleteMsg() *websocket.PreparedMessage
+	UpsertMsg() UpsertMsg
+	DeleteMsg() DeleteMsg
 }
 
 // Event contains the old and new version a Msg.  Inserts will have Before==nil, deletions will
 // have After==nil.
 type Event[T Msg] struct {
 	Before *T `json:"before"`
 	After  *T `json:"after"`
+
+	upsertCache PreparableMessage
+	deleteCache PreparableMessage
```

Comment:
Wait... the cache should be a prepared message, not a preparable message.

The idea of the cache is that you do the computationally expensive thing (marshal an object into json, possibly gzip compression and whatever else happens) only once, even if you have ten thousand subscribers.

Right now it looks like the preparefunc is being called in WriteAll, which means you'd be generating the json ten thousand times, instead of one time.